### PR TITLE
Use preview tab for diff file views

### PR DESCRIPTION
**Pull request by sheezaaziz**
Fixes https://github.com/jupyterlab/jupyterlab-git/issues/620

### CODE REVIEW

File: src/commandsAndMenu.tsx [hunk 1]
```diff
@@ -558,6 +562,34 @@ export function addCommands(
...
+                : null;
+            const tabTitle = tab.querySelector<HTMLElement>(
+              '.lm-TabBar-tabLabel'
+            );
+            tabTitle.style.fontStyle = 'italic';
+            tabTitle.addEventListener('mouseup', () => {
+              tabTitle.style.fontStyle = 'normal';
+              if (PreviewMainAreaWidget.previewWidget === diffWidget) {
+                PreviewMainAreaWidget.previewWidget = null;
+              }
```
Comment (fcollonval) [inline src/commandsAndMenu.tsx:589]: Could you create a static method in `PreviewMainAreaWidget` like `pin(widget: Widget)` that will do that?

File: src/components/diff/PreviewMainAreaWidget.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { Message } from '@lumino/messaging';
+import { Widget } from '@lumino/widgets';
+
+export class PreviewMainAreaWidget<
+  T extends Widget = Widget
+> extends MainAreaWidget {
+  /**
+   * Handle on the preview widget
+   */
+  static previewWidget: PreviewMainAreaWidget | null = null;
```
Comment (fcollonval) [inline src/components/diff/PreviewMainAreaWidget.ts:11]: Let's make this protected

```suggestion
  protected static previewWidget: PreviewMainAreaWidget | null = null;
```

File: src/commandsAndMenu.tsx [hunk 3]
```diff
@@ -558,6 +562,34 @@ export function addCommands(
...
+
+          let tabPosition = -1;
+          const tabBar = find(dockPanel.tabBars(), bar => {
+            tabPosition = bar.titles.indexOf(diffWidget.title);
+            console.log('2', tabPosition);
+            return tabPosition !== -1;
+          });
+
+          // Get the tab
+          setTimeout(() => {
```
Comment (fcollonval) [inline src/commandsAndMenu.tsx:576]: Let's add some explanation about the need for `setTimeout`

```suggestion
          // We need to wait for the tab node to be inserted in the DOM
          setTimeout(() => {
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @sheezaaziz I have a few suggestions.

File: style/diff-nb.css [hunk 4]
```diff
@@ -143,6 +143,10 @@
 	border: none;
 }
 
+.jp-git-tab-mod-preview {
+  font-style: italic;
+}
+
```
Comment (fcollonval) [inline style/diff-nb.css:149]: Would you mind placing this block at the end of `style/base.css`.

File: src/components/diff/PreviewMainAreaWidget.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        tabPosition >= 0 ? tabBar.contentNode.children[tabPosition] : null;
+      const tabTitle = tab.querySelector<HTMLElement>('.lm-TabBar-tabLabel');
+
+      tabTitle.classList.add('jp-git-tab-mod-preview');
+      tabTitle.onclick = () => {
+        tabTitle.classList.remove('jp-git-tab-mod-preview');
+        if (PreviewMainAreaWidget.previewWidget === diffWidget) {
+          PreviewMainAreaWidget.previewWidget = null;
+        }
+      };
```
Comment (fcollonval) [inline src/components/diff/PreviewMainAreaWidget.ts:52]: Almost correct, we also want to remove the click listener once it has been triggered or when the diff view is disposed:

```suggestion
      const onClick = () => {
        tabTitle.classList.remove('jp-git-tab-mod-preview');
        tabTitle.removeEventListener('click', onClick, true);
        if (PreviewMainAreaWidget.previewWidget === diffWidget) {
          PreviewMainAreaWidget.previewWidget = null;
        }
      };
      
      tabTitle.addEventListener('click', onClick, true);
      diffWidget.disposed.connect(() => {
        tabTitle.removeEventListener('click', onClick, true);
      });
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thank @sheezaaziz 

Two small things and we are good to go.

**Review by fcollonval — COMMENTED** [review]
Thanks a lot @sheezaaziz 

I tested and look again at the code. This is great. I rerun the tests failing hoping it is a random error.

**Review by fcollonval — APPROVED** [review]

**Comment by fcollonval** [comment]
https://github.com/all-contributors please add @sheezaaziz for code

**Comment by fcollonval** [comment]
@all-contributors please add @sheezaaziz for code

**Comment by krassowski** [comment]
Great work here, it is a nice feature to have!
- Could we add a setting to disable this behaviour? It seems that https://github.com/jupyterlab/jupyterlab-git/issues/620 proposed adding this as an opt-in setting.
- When I move the diff to another spit panel (say to the right):
   - the italic style in tab bar disappears (probably the class does not stick after reattaching in another panel)
   - if I open another diff the old diff in right panel gets closed and reopened in the left panel. I find this annoying, I guess it was not intended.
- I had no idea what italic style means before looking at the docs; it would be helpful if there was an explanation in the title, or maybe even a pin icon; a pin icon could be implemented as unicode prefix (📌) added to title

**Pull request merged**